go-controller/pkg/libovsdbops/router.go

```diff
@@ -7,116 +7,92 @@ import (
+		return nil, err
 	}
 
-	router.UUID = routers[0].UUID
-	return &routers[0], nil
+	return found[0], nil
 }
 
-func AddLoadBalancersToRouterOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, router *nbdb.LogicalRouter, lbs ...*nbdb.LoadBalancer) ([]libovsdb.Operation, error) {
-	if ops == nil {
-		ops = []libovsdb.Operation{}
-	}
-	if len(lbs) == 0 {
-		return ops, nil
-	}
+func FindLogicalRoutersWithPredicate(nbClient libovsdbclient.Client, p logicalRouterPredicate) ([]*nbdb.LogicalRouter, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
+	defer cancel()
+	found := []*nbdb.LogicalRouter{}
+	err := nbClient.WhereCache(p).List(ctx, &found)
+	return found, err
+}
 
-	_, err := findRouter(nbClient, router)
-	if err != nil {
-		return nil, fmt.Errorf("unable to find router %q: %w", router.Name, err)
-	}
+// LB OPs
 
-	lbUUIDs := make([]string, 0, len(lbs))
+// AddLoadBalancersToLogicalRouterOps adds the provided load balancers to the
+// provided logical router and returns the corresponding ops
+func AddLoadBalancersToLogicalRouterOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, router *nbdb.LogicalRouter, lbs ...*nbdb.LoadBalancer) ([]libovsdb.Operation, error) {
+	originalLBs := router.LoadBalancer
+	router.LoadBalancer = make([]string, 0, len(lbs))
 	for _, lb := range lbs {
-		lbUUIDs = append(lbUUIDs, lb.UUID)
+		router.LoadBalancer = append(router.LoadBalancer, lb.UUID)
 	}
-
-	op, err := nbClient.Where(router).Mutate(router, model.Mutation{
-		Field:   &router.LoadBalancer,
-		Mutator: libovsdb.MutateOperationInsert,
-		Value:   lbUUIDs,
-	})
-	if err != nil {
-		return nil, err
+	opModel := OperationModel{
+		Name:             router.Name,
+		Model:            router,
+		ModelPredicate:   func(item *nbdb.LogicalRouter) bool { return item.Name == router.Name },
+		OnModelMutations: []interface{}{&router.LoadBalancer},
+		ErrNotFound:      true,
+		BulkOp:           false,
 	}
-	ops = append(ops, op...)
 
-	return ops, nil
+	modelClient := NewModelClient(nbClient)
+	ops, err := modelClient.CreateOrUpdateOps(ops, opModel)
+	router.LoadBalancer = originalLBs
+	return ops, err
 }
 
-func RemoveLoadBalancersFromRouterOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, router *nbdb.LogicalRouter, lbs ...*nbdb.LoadBalancer) ([]libovsdb.Operation, error) {
-	if ops == nil {
-		ops = []libovsdb.Operation{}
-	}
-	if len(lbs) == 0 {
-		return ops, nil
-	}
-
-	_, err := findRouter(nbClient, router)
-	if err != nil {
-		return nil, fmt.Errorf("unable to find router %q: %w", router.Name, err)
-	}
-
-	lbUUIDs := make([]string, 0, len(lbs))
+// RemoveLoadBalancersFromLogicalRouterOps removes the provided load balancers from the
+// provided logical router and returns the corresponding ops
+func RemoveLoadBalancersFromLogicalRouterOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, router *nbdb.LogicalRouter, lbs ...*nbdb.LoadBalancer) ([]libovsdb.Operation, error) {
```

Comment:
Hmmm not sure, they are identical yes but mostly declarative and small. When you have a tool like model client that is actually handling a lot of the stuff for you, a lot of the methods using it will look very similar. I would not like just because of that to convert this into a `passing functions around` fest. I personally don't mind similar methods as long as there is not much logic into them.
Let me know what you think.

Reply:
Fair enough. If anything, my idea was to keep these 2 and they would use a common function that did all the work and the one different operation would be provided by the caller. Give they are not that big, I take it back; let's keep them as is... sorry for the noise. ;)